source/common/router/router.cc

```diff
@@ -111,6 +111,14 @@ FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_he
     request_headers.insertEnvoyExpectedRequestTimeoutMs().value(expected_timeout);
   }
 
+  // If we've configured max_grpc_timeout, override the grpc-timeout header with
+  // the expected timeout. This ensures that the optional per try timeout is reflected
+  // in grpc-timeout, ensuring that the upstream gRPC server is aware of the actual timeout.
+  if (grpc_request && route.maxGrpcTimeout() && request_headers.GrpcTimeout()) {
```

Comment:
should we insert grpc-timeout anyway regardless whether grpc-timeout exists in the request?

Reply:
Yeah I think that's a good idea, updated